src/adapter/src/copy_to.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementations around supporting the COPY TO protocol with the dataflow layer
+
+use std::collections::BTreeSet;
+
+use mz_adapter_types::connection::ConnectionId;
+use mz_controller_types::ClusterId;
+use mz_repr::GlobalId;
+use mz_sql::session::user::User;
+
+use crate::{AdapterError, ExecuteContext, ExecuteResponse};
+
+/// A description of an active copy to from coord's perspective.
+#[derive(Debug)]
+pub struct ActiveCopyTo {
+    /// The user of the session that created the copy to.
+    pub user: User,
+    /// The connection id of the session that created the copy to.
+    pub conn_id: ConnectionId,
+    /// The context about the COPY TO statement getting executed.
+    /// Used to eventually call `ctx.retire` on.
+    pub ctx: Option<ExecuteContext>,
+    /// The cluster that the copy to is running on.
+    pub cluster_id: ClusterId,
+    /// All `GlobalId`s that the copy to's expression depends on.
+    pub depends_on: BTreeSet<GlobalId>,
+    /// Whether we are already in the process of dropping the resources related to this copy to.
+    pub dropping: bool,
+}
+
+impl ActiveCopyTo {
+    /// Processes a copy to response.
+    pub(crate) fn process_response(&mut self, result: Result<u64, anyhow::Error>) {
+        // Using an option to get an owned value after take
+        // to call `retire` on it.
+        if let Some(ctx) = self.ctx.take() {
```

Comment:
Could `process_response` instead take `self` (i.e., by value)? I think that would avoid the `Option` around `ctx`, and also make it harder to forget to remove `ActiveCopyTo` from the `active_copy_tos` map when you receive the first response.

Reply:
Oh I see. Makes sense, let me update.

Reply:
The way the code's written now, we process the response, and then drop the compute sinks (which again fetches the entry from `active_copy_tos` map), and update some metrics (which reads from `active_copy_tos` as well). So, if I consume the object in `process_response`, the sink does not get dropped (and when dropping the cluster in my test it panics). I could refactor it, but subscribe also uses these methods so does not seem worth the effort? Keeping as it is now.

Reply:
It seems to me you could push the call to `process_response` into the call to `remove_active_copy_to`, where you've got the owned reference?

```diff
diff --git a/src/adapter/src/coord/message_handler.rs b/src/adapter/src/coord/message_handler.rs
index 93719b1e8f..e8c1cbafa3 100644
--- a/src/adapter/src/coord/message_handler.rs
+++ b/src/adapter/src/coord/message_handler.rs
@@ -368,13 +368,12 @@ impl Coordinator {
             }
             ControllerResponse::CopyToResponse(sink_id, response) => {
                 if let Some(active_copy_to) = self.active_copy_tos.get_mut(&sink_id) {
-                    active_copy_to.process_response(response);
                     let csid = ComputeSinkId {
                         cluster_id: active_copy_to.cluster_id,
                         global_id: sink_id,
                     };
                     self.drop_compute_sinks([csid]);
-                    self.remove_active_copy_to(sink_id);
+                    self.remove_active_copy_to(sink_id, response);
                 }
             }
             ControllerResponse::ComputeReplicaMetrics(replica_id, new) => {
diff --git a/src/adapter/src/coord/sequencer/inner/peek.rs b/src/adapter/src/coord/sequencer/inner/peek.rs
index 0be77f4f5b..1daf528439 100644
--- a/src/adapter/src/coord/sequencer/inner/peek.rs
+++ b/src/adapter/src/coord/sequencer/inner/peek.rs
@@ -844,7 +844,7 @@ impl Coordinator {
         let active_copy_to = ActiveCopyTo {
             user: ctx.session().user().clone(),
             conn_id: ctx.session().conn_id().clone(),
-            ctx: Some(ctx),
+            ctx,
             cluster_id,
             depends_on: validity.dependency_ids.clone(),
             dropping: false,
diff --git a/src/adapter/src/coord/sql.rs b/src/adapter/src/coord/sql.rs
index c8234f8a17..4bcd1d65a3 100644
--- a/src/adapter/src/coord/sql.rs
+++ b/src/adapter/src/coord/sql.rs
@@ -270,13 +270,14 @@ impl Coordinator {
 
     /// Handle removing metadata associated with a COPY TO query.
     #[tracing::instrument(level = "debug", skip(self))]
-    pub(crate) fn remove_active_copy_to(&mut self, id: GlobalId) {
+    pub(crate) fn remove_active_copy_to(&mut self, id: GlobalId, response: Result<u64, anyhow::Error>) {
         if let Some(active_copy_to) = self.active_copy_tos.remove(&id) {
             let session_type = metrics::session_type_label_value(&active_copy_to.user);
             self.metrics
                 .active_copy_tos
                 .with_label_values(&[session_type])
                 .dec();
+            active_copy_to.process_response(response);
         }
     }
 }
diff --git a/src/adapter/src/copy_to.rs b/src/adapter/src/copy_to.rs
index 8e7ff72ad8..f5a1ed3935 100644
--- a/src/adapter/src/copy_to.rs
+++ b/src/adapter/src/copy_to.rs
@@ -27,7 +27,7 @@ pub struct ActiveCopyTo {
     pub conn_id: ConnectionId,
     /// The context about the COPY TO statement getting executed.
     /// Used to eventually call `ctx.retire` on.
-    pub ctx: Option<ExecuteContext>,
+    pub ctx: ExecuteContext,
     /// The cluster that the copy to is running on.
     pub cluster_id: ClusterId,
     /// All `GlobalId`s that the copy to's expression depends on.
@@ -38,15 +38,11 @@ pub struct ActiveCopyTo {
 
 impl ActiveCopyTo {
     /// Processes a copy to response.
-    pub(crate) fn process_response(&mut self, result: Result<u64, anyhow::Error>) {
-        // Using an option to get an owned value after take
-        // to call `retire` on it.
-        if let Some(ctx) = self.ctx.take() {
-            let response = match result {
-                Ok(n) => Ok(ExecuteResponse::Copied(n.try_into().expect("u64 to usize"))),
-                Err(error) => Err(AdapterError::Unstructured(error)),
-            };
-            let _ = ctx.retire(response);
-        }
+    pub(crate) fn process_response(self, result: Result<u64, anyhow::Error>) {
+        let response = match result {
+            Ok(n) => Ok(ExecuteResponse::Copied(n.try_into().expect("u64 to usize"))),
+            Err(error) => Err(AdapterError::Unstructured(error)),
+        };
+        let _ = self.ctx.retire(response);
     }
 }
```

I didn't actually test this, so it's possible it still panics, but it seemed like it'd let you keep the control flow more or less the same as `SUBSCRIBE`, while still taking ownership of the `ActiveCopyTo` while processing the response.

Reply:
@benesch It does not feel right to me to process the request in a remove wrapper call (where it was just removing an entry and decrementing a counter)? I guess I can rename the method from `remove_active_copy_to` to `process_active_copy_to` which also removes the entry.

Reply:
Yeah, exactly! Could even call it “process_and_remove_active_copy_to”, or “retire_active_copy_to”? I like the latter the best I think.

Reply:
updated